src/repr/src/row/collection.rs

```diff
@@ -131,42 +155,39 @@ impl RowCollection {
         Some((row, upper))
     }
 
-    /// "Sorts" the [`RowCollection`] by returning a sorted view over the collection.
-    pub fn sorted_view<F>(self, mut f: F) -> SortedRowCollection
-    where
-        F: FnMut(&RowRef, &RowRef) -> std::cmp::Ordering,
-    {
-        let mut view: Vec<_> = (0..self.metadata.len()).collect();
-        view.sort_by(|a, b| {
-            let (a_row, _) = self.get(*a).expect("index invalid?");
-            let (b_row, _) = self.get(*b).expect("index invalid?");
-            f(a_row, b_row)
-        });
+    /// "Sorts" the [`RowCollection`] by returning a sorted view over the collection. Already
+    /// sorted row collections are not sorted again.
+    fn sorted_view(
+        self,
+        sort_by: Arc<Mutex<dyn FnMut(&RowRef, &RowRef) -> Ordering + Send + Sync>>,
+    ) -> SortedRowCollection {
+        let mut view = None;
+
+        if !self.is_sorted {
+            let mut indices = (0..self.metadata.len()).collect::<Vec<_>>();
+            let mut sort_by = sort_by.lock().expect("Mutex should lock");
+            indices.sort_by(|a, b| {
+                let (left, _) = self.get(*a).expect("index invalid?");
+                let (right, _) = self.get(*b).expect("index invalid?");
+
+                sort_by(left, right)
+            });
+            view = Some(indices.into())
+        }
 
         SortedRowCollection {
             collection: self,
-            sorted_view: view.into(),
+            sorted_view: view,
         }
     }
-}
-
-impl<'a, T: IntoIterator<Item = &'a Row>> From<T> for RowCollection {
-    fn from(rows: T) -> Self {
-        let mut encoded = Vec::<u8>::new();
-        let mut metadata = Vec::<EncodedRowMetadata>::new();
 
-        for row in rows {
-            encoded.extend(row.data());
-            metadata.push(EncodedRowMetadata {
-                offset: encoded.len(),
-                diff: unsafe { NonZeroUsize::new_unchecked(1) },
-            });
-        }
+    pub fn sorted_view_cost(&self) -> usize {
+        let sorted_view_mem = self.entries().saturating_mul(size_of::<usize>());
+        self.byte_len().saturating_add(sorted_view_mem)
```

Comment:
Shifted from `relation.rs`.

Why is `self.byte_len()` added to the cost? Isn't the existing allocation used when creating the sorted view?